e3db/src/main/java/com/tozny/e3db/Client.java

```diff
@@ -1623,12 +1627,84 @@ public void run() {
-   * @param dest Destination to write the decrypted file to. If the file exists, it will be truncated.
+   * @param dest Destination to write the decrypted file to. Cannot be {@code null}. If the file exists, it will be truncated.
    * @param handleResult Handles the result of the operation. If the operation completes successfully, the
    *                     destination will hold the contents of the unencrypted file.
+   *
+   *                     <p>If an error occurs, be sure to truncate the destination file, as it may contain partially
+   *                     decrypted result.
    */
-  public void readFile(UUID recordId, File dest, ResultHandler<RecordMeta> handleResult) {
-    throw new IllegalStateException("Not yet implemented.");
+  public void readFile(final UUID recordId, File dest, final ResultHandler<RecordMeta> handleResult) {
+    try {
+      checkNotNull(recordId, "recordId");
+      checkNotNull(dest, "dest");
+      final File absDest = dest.getAbsoluteFile();
+      if(! absDest.canWrite())
+        throw new IOException("Can't write to " + dest);
+
+      onBackground(new Runnable() {
+        @Override
+        public void run() {
+          try {
+            final retrofit2.Response<ResponseBody> response = storageClient.readFile(recordId.toString()).execute();
+            if (response.code() != 200) {
+              if (response.code() == 404)
+                uiError(handleResult, new E3DBNotFoundException(recordId));
+              else
+                uiError(handleResult, E3DBException.find(response.code(), response.message()));
+              return;
+            }
+
+            JsonNode result = mapper.readTree(response.body().string());
+            JsonNode meta = result.get("meta");
+            EAKEntry eak = getEAK(UUID.fromString(meta.get("writer_id").asText()),
+                UUID.fromString(meta.get("user_id").asText()),
+                clientId, meta.get("type").asText());
+
+            if(eak == null) {
+              uiError(handleResult, new E3DBUnauthorizedException("Can't read records of type " + meta.get("type").asText()));
+              return;
+            }
+
+            JsonNode fileMeta = meta.get("file_meta");
+            String signedUrl = fileMeta.get("file_url").asText();
+            Response getFileResponse = anonymousClient.newCall(new Request.Builder()
+                                                                   .url(signedUrl)
+                                                                   .get()
+                                                                   .build()).execute();
+            ResponseBody fileBody = getFileResponse.body();
+            InputStream in = fileBody.byteStream();
+            File encrypted = File.createTempFile("enc-", ".bin", new File(absDest.getParent()));
+            try {
+              FileOutputStream out = new FileOutputStream(encrypted);
+              try {
+                byte[] bytes = new byte[65_536];
```

Comment:
Can BLOCK_SIZE be referenced here instead of using a literal?

Reply:
Wouldn't really mean anything - this loop just writes the file to storage as it downloads.